new: Include a sub-section about building a server image (#18)
canonical/ubuntu-for-jetson: docs/how-to/build-image.rst

.. _build-image:

Create a custom Ubuntu Server image
===================================

Prerequisites
-------------

* `Ubuntu-image`_ must be installed on a build environment running Ubuntu 20.04 (Focal Fossa) or newer. It is recommended to use Ubuntu 24.04 (Noble Numbat).

.. _Ubuntu-image: https://github.com/canonical/ubuntu-image

.. code-block:: bash

    sudo snap install --classic --channel latest/stable ubuntu-image

Create an image definition file
-------------------------------

An image definition YAML file is required to build a custom Ubuntu Server image. This file specifies the required configurations, such as the ones listed below.

> * **class**: Type of image cloud, installer, or preinstalled.

Reviewer: "Type of image: cloud, installer or preinstalled" ?

Author: Thanks, I fixed the punctuation.

> * **kernel**: By default there is just one kernel and defaults to "linux", but through this field there can be specified an alternative kernel to install in the image.

Reviewer: As a end user, I don't understand this sentence. Can I install a kernel like BSD, seL4 or Hurd? Or do we provide other kernel configurations, if so, which ones? Or should the end user provide his own kernel? ...

Reviewer 2: I agree with your comment. Also, isn't the default "linux-nvidia-tegra-jetson"?

Since we don't provide any alternative kernel configs as far as I'm aware, this field should probably always stay the default value. Maybe an alternative would be some alternative kernel published in a ppa by the ODM?

Author: Agree, I updated that field description as follows, what do you think?
```
This field specifies the kernel for installation in the image. While linux-nvidia-tegra-jetson is the standard publicly available version for Tegra platforms, it can be replaced with an alternative kernel, such as one from a PPA, if required.
```

.. _image definition: https://github.com/canonical/ubuntu-image/blob/main/internal/imagedefinition/README.rst

The `ubuntu-images`_ repository serves as the official source for Canonical’s image definitions and scripts. It is used to build various Ubuntu versions and flavors. This includes optimized versions for Jetson Tegra hardware, currently supporting Ubuntu 22.04 (Jammy Jellyfish) and 24.04 (Noble Numbat).

.. _ubuntu-images: https://git.launchpad.net/ubuntu-images

.. tabs::

   .. group-tab:: Jammy

      .. code-block:: bash

            name: ubuntu-server-tegra-jetson-arm64
            display-name: Ubuntu Server Tegra Jetson arm64
            revision: 2
            architecture: arm64
            series: jammy
            class: preinstalled
            kernel: linux-nvidia-tegra-jetson
            gadget:
              url: "https://git.launchpad.net/~canonical-foundations/snap-pc/+git/github-mirror-amd64"
              branch: classic
              type: "git"
            rootfs:
              mirror: "http://ports.ubuntu.com/ubuntu-ports/"
              pocket: updates
              components:
                - main
                - restricted

>                 # TODO should ideally not enable universe during image build; kernel is now

Reviewer: This is internal TODO, better not push it?

Reviewer 2: This is part of our official image definitions in the ubuntu-images repo

                # in main, but need to MIR nvidia-tegra-defaults and list it in a seed (see
                # LP #2083775 for a proposal to add seeds)
                - universe
                - multiverse
              seed:
                urls:
                  - "git://git.launchpad.net/~ubuntu-core-dev/ubuntu-seeds/+git/"
                branch: jammy
                names:
                  - server
                  - minimal
                  - standard
                  - cloud-image

>                   # TODO would like to list server-tegra here once available (see

Reviewer: This is internal TODO, better not push this?

                  # LP #2083775)
            customization:
              # We need to duplicate the list of components as we currently have some

>               # packages in universe and multiverse

Reviewer: Does it matter to end users?

              components:
                - main
                - restricted
                - universe
                - multiverse
              cloud-init:
                user-data: |

>                   #cloud-config

Reviewer: What is the point of this commented out line?

                  chpasswd:
                    expire: true
                    users:
                      - name: ubuntu

>                         password: ubuntu

Reviewer: Pushing a default password that way seems fishy at best, better replace it with "#TODO"

                        type: text
                meta-data: |
                  dsmode: local
                  instance_id: ubuntu-server
              extra-snaps:
                - name: snapd
                # FIXME ubuntu-image fails in prepare_image if this isn't listed: cannot

>                 # add snap "lxd" without also adding its base "core20" explicitly

Reviewer: Is this bug still alive?

                - name: core20
              extra-packages:

>                 # FIXME using extra-packages as a workaround for missing bits to make the

Reviewer: This FIXME is for us, we should not push it in the doc, should we?

      .. code-block:: bash

            name: ubuntu-server-jetson-arm64
            display-name: Ubuntu Server Jetson arm64
            revision: 1
            architecture: arm64
            series: noble
            class: preinstalled
            kernel: linux-nvidia-tegra-jetson
            gadget:
              url: "https://git.launchpad.net/~canonical-foundations/snap-pc/+git/github-mirror-amd64"
              branch: "classic"
              type: "git"
            rootfs:
              archive: ubuntu
              components:
                - main
                - restricted
                - universe
                - multiverse
              mirror: "http://ports.ubuntu.com/ubuntu-ports/"
              sources-list-deb822: true
              pocket: updates
              seed:
                urls:
                  - "https://git.launchpad.net/~ubuntu-core-dev/ubuntu-seeds/+git/"
                branch: noble
                names:
                  - server
                  - minimal
                  - standard
                  - cloud-image
            customization:
              extra-snaps:
                - name: snapd
              extra-packages:

>                 # FIXME using extra-packages as a workaround for missing bits to make the

Reviewer: This FIXME is for us, so we should not push it in the documentation

                # system bootable that should be pulled by seeds or metapackages
                # signed EFI boot chain
                - name: grub-efi-arm64-signed
                - name: shim-signed
                # for update-grub
                - name: grub2-common
                - name: nvidia-tegra-defaults
                - name: wpasupplicant
                # install Tegra firmware files
                - name: linux-firmware-nvidia-tegra
              cloud-init:
                user-data: |
                  #cloud-config
                  chpasswd:
                    expire: true
                    users:
                      - name: ubuntu

>                         password: ubuntu

Reviewer: We should not suggest people do that, and instead leave a "#TODO" password?

.. _build-image:

Create a custom Ubuntu Server image
===================================

Prerequisites
-------------

> * `Ubuntu-image`_ must be installed on a build environment running Ubuntu 20.04 (Focal Fossa) or newer. It is recommended to use Ubuntu 24.04 (Noble Numbat).

Reviewer 3: Is it (did you try to build both the jammy and noble image with it)? Otherwise I would tend to recommend using the same host Ubuntu version as the one to build.

Author: Yes, today I tested the following..

* Build a Jammy image on a Jammy host and test that it boots in a Jammy DUT.
* Build a Jammy image on a Noble host and test that it boots in a Jammy DUT.
* Build a Noble image on a Jammy host and test that it boots in a Noble DUT.
* Build a Noble image on a Noble host and test that it boots in a Noble DUT.

Author: I posted the corresponding logs in the PR description.

.. _Ubuntu-image: https://github.com/canonical/ubuntu-image

.. code-block:: bash

    sudo snap install --classic --channel latest/stable ubuntu-image

Create an image definition file
-------------------------------

> To build a custom Ubuntu Server image, there is needed an image definition YAML file, which specifies required configurations, like the ones listed below.

Reviewer 3: ```suggestion
An image definition YAML file is required to build a custom Ubuntu Server image. This file specifies the required configurations, such as the ones listed below.
```

* **class**: Type of image cloud, installer, or preinstalled.
* **kernel**: By default there is just one kernel and defaults to "linux", but through this field there can be specified an alternative kernel to install in the image.
* **gadget**: Boot assets of an image.
* **customization**: Features such as particular snaps and packages that will come installed in the image.
* **artifacts**: Artifacts to create, including (but not limited to) the actual images, and manifest files.

For more details about each field the `image definition`_ documentation can be consulted.

.. _image definition: https://github.com/canonical/ubuntu-image/blob/main/internal/imagedefinition/README.rst

> This is an `example of the image definition`_ file used to build the preinstalled images for Jetson devices.

Reviewer 3: I would rather present in a dedicated section the reference image definitions used to produce our released Ubuntu images, and making use of the tab group feature as per https://github.com/canonical/ubuntu-for-jetson/pull/17

.. _build-image:

Create a custom Ubuntu Server image
===================================

Prerequisites
-------------

> * `Ubuntu-image`_ must be installed on a build environment running Ubuntu 20.04 (Focal Fossa) or newer. It is recommended to use Ubuntu 24.04 (Noble Numbat).

Reviewer 3: ```suggestion
* `Ubuntu-image`_ is the tool used by Canonical to build official Ubuntu images. It must be installed on a build environment running Ubuntu 20.04 (Focal Fossa) or newer. It is recommended to use Ubuntu 24.04 (Noble Numbat).
```

.. _Ubuntu-image: https://github.com/canonical/ubuntu-image

.. code-block:: bash

    sudo snap install --classic --channel latest/stable ubuntu-image

Create an image definition file
-------------------------------

An image definition YAML file is required to build a custom Ubuntu Server image. This file specifies the required configurations, such as the ones listed below.

> * **class**: Type of image: cloud, installer or preinstalled.

Reviewer 3: ```suggestion
* **class**: Defines the type of image, such as `cloud`, `installer` or `preinstalled` (Canonical's certified Server images are preinstalled).
```

> * **kernel**: This field specifies the kernel for installation in the image. While "linux-nvidia-tegra-jetson" is the standard publicly available version for Tegra platforms, it can be replaced with an alternative kernel, such as one from a PPA, if required.

Reviewer 3: ```suggestion
* **kernel**: Specifies the preinstalled kernel in the image. The official Ubuntu kernel for Jetson is "linux-nvidia-tegra-jetson", but it can be replaced with an alternative custom kernel, for instance hosted in a PPA for development purpose.
```

* **gadget**: Boot assets of an image.
* **customization**: Features such as particular snaps and packages that will come installed in the image.
* **artifacts**: Artifacts to create, including (but not limited to) the actual images, and manifest files.

> For more details about each field, the `image definition`_ documentation can be consulted.

Reviewer 3: ```suggestion
For more details about each field, refer to the `image definition`_ documentation.
```

.. _image definition: https://github.com/canonical/ubuntu-image/blob/main/internal/imagedefinition/README.rst

The `ubuntu-images`_ repository serves as the official source for Canonical’s image definitions and scripts. It is used to build various Ubuntu versions and flavors. This includes optimized versions for Jetson Tegra hardware, currently supporting Ubuntu 22.04 (Jammy Jellyfish) and 24.04 (Noble Numbat).

.. _ubuntu-images: https://git.launchpad.net/ubuntu-images

Ubuntu certified Image definition files
---------------------------------------

> Provided below are the image definition files utilized in the generation of Ubuntu Server images certified by Canonical.

Reviewer 3: Nitpick:

```suggestion
The following command allows to check the definition files used to generate the Ubuntu Server image certified by Canonical:
```

.. _Certified Ubuntu Server Jammy image definition: https://git.launchpad.net/ubuntu-images/tree/ubuntu-server-tegra-jetson.yaml?h=jammy
.. _Certified Ubuntu Server Noble image definition: https://git.launchpad.net/ubuntu-images/tree/ubuntu-server-tegra-jetson.yaml?h=noble

.. tabs::

   .. group-tab:: Jammy

      `Certified Ubuntu Server Jammy image definition`_

   .. group-tab:: Noble

      `Certified Ubuntu Server Noble image definition`_

Build the Classic Server image
------------------------------

> Once having the corresponding image definition file, the server image can be built through the following command.

Reviewer 3: ```suggestion
Once the corresponding image definition file is defined, the server image can be built through the following command.
```
